test/lib/util/aws/cloudfront_signed_url_test.exs

```diff
@@ -0,0 +1,96 @@
+  @signature_with_query3   "BzTb7kOa1P2sp2lO~sO4j17b7UZpxgDXIvR4ET5FFa0rwu9Z5u9ubC09EGjHn9r-k~pxWu-KJyvr--f0MDNtUGWR5fgOb74iobt4GP9-NsS9WRu3JdhFZXHqAGkPF9TtSXrP8qMtcsBJ~F0SWeWevdhU19FMPH4hIe9ff0m2J29BIFZUe9~fyWh3daMQ9aC7Zf84G3V00FW~v21ggcRhDgkLdN6eUCm-cayzs6kZhr0V6PABjw0Ve9hJbAHC-B2PbTwfFVkiLn0B4dQtgwYC9NHhvWzil7xGBhJvAX7CPJJNImi~A1x8FdigTgMUXg4fwAapj0tk4FmkFu-NHI-AQg__"
+
+  defp assert_query_params(query_params, expected_signature) do
+    [
+      {"Expires"    , expires    },
+      {"Signature"  , signature  },
+      {"Key-Pair-Id", key_pair_id},
+    ] = query_params
+    assert expires     == Integer.to_string(@expires_in_seconds)
+    assert signature   == expected_signature
+    assert key_pair_id == @key_pair_id
+  end
+
+  defp assert_base_url_and_get_query_params(url) do
+    %URI{host: host, path: path, query: query, scheme: scheme} = URI.parse(url)
+    assert "#{scheme}://#{host}#{path}" == URI.encode(@resource_url)
+    URI.query_decoder(query) |> Enum.to_list()
+  end
+
+  setup do
+    lifetime = 60
+    :meck.expect(System, :system_time, fn :second -> @expires_in_seconds - lifetime end)
+    on_exit(fn -> :meck.unload() end)
+    %{lifetime: lifetime}
+  end
+
+  test "should generate a signed URL for URL without query", %{lifetime: lifetime} do
+    signed_url = CloudfrontSignedUrl.generate_signed_url(@resource_url, lifetime, @key_pair_id, @private_key)
+    query_params = assert_base_url_and_get_query_params(signed_url)
+    assert_query_params(query_params, @signature_without_query)
+  end
+
+  test "should generate a signed URL for URL with query", %{lifetime: lifetime} do
+    [{"?", @signature_with_query1}, {"?hello=world", @signature_with_query2}, {~s/?hello="日本"&/, @signature_with_query3}]
+    |> Enum.each(fn {query, signature} ->
+      resource_url = @resource_url <> query
+      signed_url = CloudfrontSignedUrl.generate_signed_url(resource_url, lifetime, @key_pair_id, @private_key)
+      assert String.starts_with?(signed_url, URI.encode(resource_url) <> "&Expires=")
+      query_params = assert_base_url_and_get_query_params(signed_url)
+      query_params |> Enum.reverse() |> Enum.take(3) |> Enum.reverse() |> assert_query_params(signature)
+    end)
+  end
+
+  test "should generate a signed URL for encoded URL", %{lifetime: lifetime} do
+    [{"?", @signature_with_query1}, {"?hello=world", @signature_with_query2}, {~s/?hello="日本"&/, @signature_with_query3}]
+    |> Enum.each(fn {query, signature} ->
+      resource_url = URI.encode(@resource_url <> query)
+      signed_url = CloudfrontSignedUrl.generate_signed_url(resource_url, lifetime, @key_pair_id, @private_key, true)
+      assert String.starts_with?(signed_url, resource_url <> "&Expires=")
+      query_params = assert_base_url_and_get_query_params(signed_url)
+      query_params |> Enum.reverse() |> Enum.take(3) |> Enum.reverse() |> assert_query_params(signature)
```

Comment:
```suggestion
      query_params |> Enum.take(-3) |> assert_query_params(signature)
```
or
```ex
assert_base_url_and_get_query_params(signed_url)
|> Enum.take(-3)
|> assert_query_params(signature)
```

Reply:
It looks very smart!
Thank you.
I fixed in https://github.com/access-company/antikythera/pull/100/commits/ff031d728170f0e6818b931ea7200e6a0d815f86